tests/test_database.py

```diff
@@ -0,0 +1,333 @@
+
+
+def test_registration(session, models_with_kwargs):
+    models = models_with_kwargs.models
+    new_app = FastAPI()
+
+    # assert that this application has no registered routes
+    assert len(registered_routes(new_app)) == 0
+
+    def get_session():
+        yield session
+
+    # register routes for this application
+    abstractions.register_endpoints(api=new_app, get_session=get_session, models=models)
+
+    # assert that this application now has five registered routes
+    routes = registered_routes(new_app)
+
+    assert len(routes) == 5
+
+    expected_names = ("_create", "_read_range", "_read_single", "_update", "_delete")
+    # Check names
+    for r in routes:
+        assert r.name in expected_names
+    # Check routes
+    for r in routes:
+        if r.name in ("_create", "_read_range"):
+            assert r.path == models.path
+        elif r.name in ("_read_single", "_update", "_delete"):
+            assert r.path == f"{models.path}{{id}}"
+    # Check response models
+    for r in routes:
+        if r.name in ("_create", "_read_single", "_update"):
+            assert r.response_model == models.response
+        elif r.name == "_read_range":
+            assert r.response_model == List[models.response]
+        elif r.name == "_delete":
+            assert r.response_model is None
+
+
+# Test create ---------------------------------------------------------------------------
+
+
+class TestCreate(CreateFixtures):
+    """Container for tests of database entry creation and associated failure modes"""
+
+    @staticmethod
+    def get_error(func: Callable, failure_mode: str):
+        interface = get_interface(func)
+        errors = dict(db=IntegrityError, abstraction=ValidationError, client=HTTPError,)
+        cli_error = _MissingFieldError if failure_mode == "incomplete" else _StrTypeError
+        errors.update(dict(cli=cli_error))
+        return errors[interface]
+
+    @staticmethod
+    def evaluate_data(request: dict, data: dict, blank_opts: Optional[List] = None):
+        for k in request.keys():
+            if isinstance(data[k], datetime):
+                assert data[k] == parse_to_datetime(request[k])
+            elif (
+                # Pydantic requires a "Z"-terminated timestamp, but FastAPI responds without the "Z"
+                isinstance(data[k], str)
+                and isinstance(request[k], str)
+                and any([s.endswith("Z") for s in (data[k], request[k])])
+                and not all([s.endswith("Z") for s in (data[k], request[k])])
+            ):
+                assert add_z(data[k]) == add_z(request[k])
+            else:
+                assert data[k] == request[k]
+        if blank_opts:
+            for k in blank_opts:
+                assert data[k] is None
+        assert data["id"] is not None
+
+    def test_create(
+        self,
+        session: Session,
+        model_to_create: ModelFixture,
+        http_server: str,
+        create_func: Callable,
```

Comment:
Please help me understand where `create_func` gets defined and passed to `test_create`.

Reply:
Ok, I see that this is a fixture.

Reply:
Removed in https://github.com/pangeo-forge/pangeo-forge-orchestrator/pull/6/commits/72a24caa6252c387d1f9e2597be224ddaec07cac